react-renderer/src/Evaluator.ts

```diff
@@ -339,10 +343,12 @@ export const resolvePath = (
                 string
               ], // TODO: check if this is true
             };
-            varyingVal = varyingMap?.get(propPath);
+            varyingVal = varyingMap?.get(JSON.stringify(propPath));
             if (varyingVal) {
-              return { tag: "FloatV", contents: varyingVal[1] };
-            } else return tf ? toTF(p.contents) : p.contents;
+              return { tag: "FloatV", contents: varyingVal };
+            } else {
+              return autodiff ? differentiableValue(p.contents) : p.contents;
```

Comment:
I like this approach. I wonder if tf.js might be somehow faster at evaluating the forward pass though? (i.e. maybe everything can be a tensor but not necessarily a variable) It also feels inefficient to pack one number per tensor. Could be performance opt for later

Reply:
I wanted to minimize the use of tensors and intentionally left the evaluator as close to the plain JS version as possible, but I’m happy to chat more about opt a bit later